Reuse MappingMetadata instances in Metadata class.

martijnvg opened the pull request:
Hash the mapping source of a MappingMetadata instance and then
cache it in Metadata class. A mapping with the same hash
will use a cached MappingMetadata instance. This can
significantly reduce the number of MappingMetadata instances
for data streams and index patterns.

Idea originated from #69772, but just focusses on the jvm heap memory savings.
And hashes the mapping instead of assigning it an uuid.

Relates to #77466

martijnvg committed:
Reuse MappingMetadata instances in Metadata class.

Hash the mapping source of a MappingMetadata instance and then
cache it in Metadata class. A mapping with the same hash
will use a cached MappingMetadata instance. This can
significantly reduce the number of MappingMetadata instances
for data streams and index patterns.

Relates to #77466

martijnvg committed:
rename

martijnvg commented:
Locally I have been testing this improvement with the script at the end of this comment. This script creates a data stream and rolls it over concurrently (1000 rollovers). With this change, the number of `MappingMetadata` instances is 3 (running the script in en empty cluster) and without this change over 2000 instances.

```python
import threading
from elasticsearch import Elasticsearch

ds_base_name = 'logs-mysql-'
es = Elasticsearch(['http://localhost:9200'], http_auth=('elastic-admin', 'elastic-password'), timeout=600)

def rollover_test(ds_name):
    es.indices.delete_data_stream(name=ds_name, ignore=[400, 404])
    es.indices.create_data_stream(name=ds_name)
    for i in range(200):
        es.indices.rollover(alias=ds_name,wait_for_active_shards=0)


if __name__ == "__main__":
    es.cluster.put_settings(body={"persistent": {"cluster.max_shards_per_node": 100000}})
    es.indices.put_index_template(name='my-template', body={
        "index_patterns": ["logs-*-*"],
        "priority": 200,
        "data_stream": {},
        "composed_of": [
            "logs-mappings",
            "data-streams-mappings"
        ],
        "template": {
            "settings": {
                "index.number_of_replicas": 0
            }
        },
        "allow_auto_create": True
    })

    for i in range(5):
        t = threading.Thread(target=rollover_test, args=(ds_base_name + str(i),))
        t.start()
```

martijnvg committed:
Merge remote-tracking branch 'es/master' into reuse_mapping_metadata

martijnvg committed:
missed a few places where mapping reuse was performed,
causing inconsistencies.

martijnvg committed:
cleanup

martijnvg committed:
more cleanup

martijnvg committed:
moar cleanup

martijnvg committed:
Merge remote-tracking branch 'es/master' into reuse_mapping_metadata

martijnvg committed:
adjust cleanup logic

martijnvg committed:
added unit test

martijnvg committed:
Merge remote-tracking branch 'es/master' into reuse_mapping_metadata

martijnvg committed:
undo unrelated change

martijnvg committed:
digest can be null in case of older clusters

martijnvg committed:
fixed test

martijnvg committed:
fixed npe in mixed clusters

martijnvg committed:
compute digest in bwc case

martijnvg committed:
iter

elasticmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

martijnvg commented on server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:
```diff
@@ -1911,6 +1935,81 @@ public static Metadata fromXContent(XContentParser parser) throws IOException {
+            }
+        }
+
+        private void cleanupUnusedEntry(IndexMetadata previous) {
```
Because the cache is passed down to new `Metadata` instances, a cleanup mechanism is needed.
This approach with counting the times a `MappignMetata` instance is used and removing entries when there are no usages seems cheaper compared to checking each metadata is build whether all entries are used and then removed the unused entries. But this approach is also more complex. I'm also okay to change the cleanup logic to be simpler and checking each cache entry whether it is used in any of the indices and if unused remove.

original-brownbear replied:
If the simpler version doesn't come with problematic runtime overhead it might be fine? It seems like it should be cheap now that we can cheaply compare mappings?

martijnvg replied:
I think that is true. Also I find this implementation a bit fragile. I will change the cleanup implementation to just check for each cache entry whether it is used by any IndexMetadata.

martijnvg replied:
Implemented: 15b05afb51274069320f1db6ae29f62b95e91348

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java:
```diff
@@ -41,36 +45,53 @@
+        String mappingAsString = Strings.toString((builder, params) -> builder.mapContents(mapping));
         try {
-            this.source = new CompressedXContent((builder, params) -> builder.mapContents(mapping));
+            this.source = new CompressedXContent(mappingAsString);
```
Can't we use `this.source = new CompressedXContent((builder, params) -> builder.mapContents(mapping));` here still and compute the digest as we compress for example since we stream the bytes in that case anyway?

original-brownbear replied:
On that note, I wonder if we should just make the `Sha256` value a part of the `CompressedXContent`? I see how we do some magic here to make sure we build a deterministic version based on a map, but we could have that by just adding a static constructor to this thing? For BwC we can compute the CRC or SHA1 dynamically where/if needed I guess.
Also, that would allow us to make the `equals` in `CompressedXContent` almost free by just turning it into a string comparison :) WDYT?

martijnvg replied:
I like this idea and in fact I did think about this while working on this. I will try this out and see whether it is not too difficult.

martijnvg replied:
I did the first part of this: c7705231c9285e75c809619bb80bc0176c131657
I will work on making the sha256 really part of of `CompressedXContent`, just like crc32 checksum.
Just to double check, do we want to keep the `crc32` checksum? I think so.

original-brownbear replied:
> Just to double check, do we want to keep the crc32 checksum? I think so.

Not sure tbh. It seems we're only using it to speed up comparisons and as a `hashCode`. Both of which we can have from the sha256 value can't we?

martijnvg replied:
Right, I thought there was another usage of the crc32, but it only used for hashcode and checking consistency in assertions. Both can be done with sha256 too. I will replace crc32 completely with sha256.

martijnvg replied:
I pushed: 1961453aa16caa9468f541b975a047a15ae56e01

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java:
```diff
@@ -88,13 +109,12 @@ public static void writeMappingMetadata(StreamOutput out, ImmutableOpenMap<Strin
-            Map<String, Object> routingNode = (Map<String, Object>) withoutType.get("_routing");
-            for (Map.Entry<String, Object> entry : routingNode.entrySet()) {
-                String fieldName = entry.getKey();
+            Map<?, ?> routingNode = (Map<?, ?>) withoutType.get("_routing");
```
Why can't we have `String` keys here anymore?

martijnvg replied:
I was a bit annoyed with the `SuppressWarnings("unchecked")` and so removed it. I can undo this change and make this change in another pr.

original-brownbear replied:
Maybe, this one is a little too impactful to have additional noise in it I think :)

martijnvg replied:
Ok, I will undo this bit here in the pr.

martijnvg replied:
Removed this change: de25c023076ed9d533cd785f60bdeaad3bbd4368

original-brownbear reviewed (commented):
Thanks Martijn, this looks really nice and should be pretty impactful. I just have a couple of questions, suggestions on the plumbing around this mostly :) Happy to discuss this on another channel if that's quicker as well.

martijnvg committed:
Merge remote-tracking branch 'es/master' into reuse_mapping_metadata

martijnvg committed:
Simplify purge logic in favour for a bit higher runtime costs

martijnvg committed:
First step of moving sha256 to CompressedXContent, by letting it get computed
from from the compressed bytes in CompressedXContent.

martijnvg committed:
Change CompressedXContent to sha256 instead of crc32 for hashcode and consistency checking.
Also change CompressedXContent#equals() method to compare using the sha256 only.
Dedupe/cache MappingMetadata instance by teh sha256 that CompressedXContent creates.

martijnvg committed:
undo change

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:
```diff
@@ -230,7 +231,8 @@ private Metadata(
         String[] visibleClosedIndices,
-        SortedMap<String, IndexAbstraction> indicesLookup
+        SortedMap<String, IndexAbstraction> indicesLookup,
+        Map<String, MappingMetadata> cache
```
NIT: Maybe call this something mapping specific, like `mappingsByHash` or so? :) `cache` seems a little generic :)

martijnvg replied:
I've renamed this: 3bd8a655ddc506f88297f79d9339138cd198f73b

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:
```diff
@@ -1911,6 +1930,60 @@ public static Metadata fromXContent(XContentParser parser) throws IOException {
+            if (entry != null) {
+                IndexMetadata.Builder imBuilder = new IndexMetadata.Builder(indexMetadata);
+                imBuilder.putMapping(entry);
+                return imBuilder.build();
```
This builder is fairly expensive if I remember correctly with all the setting lookup and the like. I wonder if we could just make this a copy constructor that swaps out the mapping metadata only like we did elsewhere and ourselves the trouble of setting up a builder.
Also, in this case you could make it transparent easily, and not create a new instance even in case the mappings are already instance equal?

martijnvg replied:
Something like this: 4bc2229d6139e7c1e2da1a9b830ea8fc7988e63d ?
I think we need to check whether the instances are not the same and I think here instances are always different?

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:
```diff
@@ -1911,6 +1930,60 @@ public static Metadata fromXContent(XContentParser parser) throws IOException {
+
+        private void purgeMappingCache(ImmutableOpenMap<String, IndexMetadata> indices) {
+            final var iterator = cache.entrySet().iterator();
+            while (iterator.hasNext()) {
```
This seems needlessly hard :) Can't we just collect all the hash strings from the metadata into a set in one loop and then all keys from `cache` that aren't in that set? That way we're at least O(N)? :)

martijnvg replied:
Something like this: 3d2bf906b839b22330b047cc0441e4b3dfb0e040 ?

original-brownbear replied:
++ perfect

original-brownbear commented on server/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java:
```diff
@@ -41,43 +43,43 @@
             throw new Error(bogus);
         }
-        return (int) crc32.getValue();
+        return MessageDigests.toHexString(messageDigest.digest());
```
Maybe we should base64 encode these instead to save some bytes?

martijnvg replied:
I've changed this to use base64: 17f00d92ee1dc481579188491ff8e53c8c7bd733

original-brownbear reviewed (commented):
Thanks Martijn, this looks really nice (and surprisingly simple as well :)).
Only a couple comments on the details that I missed on the first round, other than that, this looks good to go.

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:
```diff
@@ -1911,6 +1930,60 @@ public static Metadata fromXContent(XContentParser parser) throws IOException {
             return builder.build();
         }
+
+        private IndexMetadata reuseMappings(IndexMetadata indexMetadata) {
```
One more thing, `cache` (or whatever we might rename it to) should have a little bit of JavaDoc that roughly explains what's going on here.

martijnvg replied:
This was done as part of the rename commit: 3bd8a655ddc506f88297f79d9339138cd198f73b

martijnvg committed:
Merge remote-tracking branch 'es/master' into reuse_mapping_metadata

martijnvg committed:
rename

martijnvg committed:
undo unrelated change

original-brownbear commented on server/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java:
```diff
@@ -43,43 +43,43 @@
     private static final ThreadLocal<InflaterAndBuffer> inflater1 = ThreadLocal.withInitial(InflaterAndBuffer::new);
     private static final ThreadLocal<InflaterAndBuffer> inflater2 = ThreadLocal.withInitial(InflaterAndBuffer::new);
```
@martijnvg I just realized, we can probably remove all of this infrastructure here now that we have efficient `equals`? I added this only to speed up comparison a while back, but now it seems irrelevant and can be simplified?

martijnvg replied:
Yes, I've removed inflater2 and equalsWhenUncompressed() method: c05791c781e3ca27a56a190bb265a49770da8db3

`inflater1 ` and a few other things are still in use by production code.

martijnvg committed:
base64 encode sha256 digest

martijnvg committed:
removed equalsWhenUncompressed() method

martijnvg committed:
make cleanup unused entries less hard

martijnvg committed:
replace builder with a new copy constructor that only changes the mapping metadata instance

martijnvg committed:
Merge remote-tracking branch 'es/master' into reuse_mapping_metadata

martijnvg commented:
@elasticmachine run elasticsearch-ci/part-2

martijnvg committed:
Merge remote-tracking branch 'es/master' into reuse_mapping_metadata

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -561,6 +561,44 @@ private IndexMetadata(
         assert numberOfShards * routingFactor == routingNumShards : routingNumShards + " must be a multiple of " + numberOfShards;
     }
 
+    IndexMetadata(IndexMetadata indexMetadata, MappingMetadata mapping) {
```
Could we maybe make this just a method:

`IndexMetadata withMappingMetadata(MappingMetadata mapping)` to give us an obvious copy constructor. I also wonder, even though it's slightly more code, maybe it's better if this would just invoke the existing constructor so we don't have two constructors to maintain for this thing (given how additional fields are incoming from settings moving to fields in this class)?

martijnvg replied:
👍 I agree this is cleaner. I've made this change: 7d092821070b0983e4280a07be21251647e40636

original-brownbear reviewed (approved):
LGTM, I think I'd like the index-metadata constructor to be set up differently just to enable follow-ups but it's your call, I can deal with that myself in a follow-up when I work on it again :)
Thanks so much for the iterations on this one!

martijnvg committed:
Merge remote-tracking branch 'es/master' into reuse_mapping_metadata

martijnvg committed:
Turn copy constructor into method (withMappingMetadata(...))

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.0  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 80348`
